### ISSUE TITLE
switch-buffer-next does not cycle over no-save buffers

### ISSUE THREAD

**Issue by Ambrevar** [atlas-engineer/nyxt#1794]
Currently we flatten the buffer tree following the inter-buffer relationships from the Global History Tree.

Problem: we didn't take into account that some buffers like no-save buffers may not belong to the same global history tree!
Some buffers may not even have any history at all.

Fix suggestion: go throught the list of buffers and fetch their global history, then flatten each buffer tree one after the other.  For buffers without buffer tree, list them in order of appearance at then end.

Thoughts?

**Comment by Ambrevar** [comment]
OK, I'll send a patch later then.

**Comment by Ambrevar** [comment]
I could not reproduce, it looks like we do take multiple histories into account.  Or am I missing something?

Anyways, I've spotted a bug in how we call `get-data`.  This API is broken :(  Until we rewrite the data-storage library, I've sent a fix with 14b0abc6c031050cfb814b8b84d828fa87c09956, should be good for now.

**Comment by Ambrevar** [comment]
Explanation: in `buffer-siblings`, we _do_ handle all "top level owners", so all buffers that don't have a creator.  

We also compute the owner of a BUFFER from the current-buffer history.  So if BUFFER has another history, the owner won't be found and will be considered a top-level buffer.  

This may not be the right code nor the right behaviour, but it allows for cycling through all buffers it seems, which is good enough considering all the other buffer switching options that we have.

We can fine-tune this in the future when the data-storage library is rewritten.

**Comment by aadcg** [comment]
@Ambrevar still a priority?

**Comment by aartaka** [comment]
I'd say yes, because it's a significant expectations violation.

**Comment by aadcg** [comment]
Is that violation felt with Nyxt's defaults? If it's not, then I'd say it's not a priority.

**Comment by aartaka** [comment]
> Is that violation felt with Nyxt's defaults? If it's not, then I'd say it's not a priority.

What do you mean? Nosave buffers are integral to Nyxt. So yes, it's bad that they are not cycled through like other buffers. Except for the data saving, they are regular buffers and should be treated as ones.

**Comment by aadcg** [comment]
I was wondering whether this bug is observed when global history is disabled (the default). I didn't make serious effort to understand the bug. My goal is to filter out the priority fixes for 3.0.

**Comment by aartaka** [comment]
No problem, now we're certain it's a priority one <3

**Comment by aadcg** [comment]
Any taker? @Ambrevar @aartaka

**Comment by Ambrevar** [comment]
I can do it.  Probably not this week though, so if someone wants to take this on this week, please go ahead (but signal it here).

The challenge is to simplify and generalize `buffer-siblings` and friends so that we guarantee that the _whole_ buffer tree is browsed, respecting the GHT but also including buffers that are not GHT owners.

**Comment by jmercouris** [comment]
It is a problem, but I don't think I would delay the release of 3.0.0 for this. It won't break the user experience. In some ways one could even argue that this is a *feature* by not cycling through them with C-tab.

**Comment by aadcg** [comment]
No-save buffers deleted in c3e5d897c3752774668660343bb8216d200e1dc8.

Related to GHT, see #3598.
